engine/src/witness/hub/hub_deposits.rs

```diff
@@ -0,0 +1,333 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use cf_primitives::{
+	EpochIndex, PolkadotBlockNumber, ASSETHUB_USDC_ASSET_ID, ASSETHUB_USDT_ASSET_ID,
+};
+use futures_core::Future;
+use pallet_cf_ingress_egress::{DepositChannelDetails, DepositWitness};
+use state_chain_runtime::AssethubInstance;
+
+use super::super::common::chunked_chain_source::chunked_by_vault::{
+	builder::ChunkedByVaultBuilder, ChunkedByVault,
+};
+use crate::{
+	dot::PolkadotHash,
+	witness::{
+		common::{
+			chunked_chain_source::chunked_by_vault::deposit_addresses::Addresses,
+			RuntimeCallHasChain, RuntimeHasChain,
+		},
+		hub::EventWrapper,
+	},
+};
+use cf_chains::{assets::hub::Asset, dot::PolkadotAccountId, Assethub};
+use subxt::events::Phase;
+
+impl<Inner: ChunkedByVault> ChunkedByVaultBuilder<Inner> {
+	pub fn hub_deposits<ProcessCall, ProcessingFut>(
+		self,
+		process_call: ProcessCall,
+	) -> ChunkedByVaultBuilder<
+		impl ChunkedByVault<
+			Index = PolkadotBlockNumber,
+			Hash = PolkadotHash,
+			Data = Vec<(Phase, EventWrapper)>,
+			Chain = Assethub,
+			ExtraInfo = PolkadotAccountId,
+			ExtraHistoricInfo = (),
+		>,
+	>
+	where
+		Inner: ChunkedByVault<
+			Index = PolkadotBlockNumber,
+			Hash = PolkadotHash,
+			Data = (Vec<(Phase, EventWrapper)>, Addresses<Inner>),
+			Chain = Assethub,
+			ExtraInfo = PolkadotAccountId,
+			ExtraHistoricInfo = (),
+		>,
+		ProcessCall: Fn(state_chain_runtime::RuntimeCall, EpochIndex) -> ProcessingFut
+			+ Send
+			+ Sync
+			+ Clone
+			+ 'static,
+		ProcessingFut: Future<Output = ()> + Send + 'static,
+		state_chain_runtime::Runtime: RuntimeHasChain<Inner::Chain>,
+		state_chain_runtime::RuntimeCall:
+			RuntimeCallHasChain<state_chain_runtime::Runtime, Inner::Chain>,
+	{
+		self.then(move |epoch, header| {
+			let process_call = process_call.clone();
+			async move {
+				let (events, addresses_and_details) = header.data;
+
+				let addresses = address_and_details_to_addresses(addresses_and_details);
+
+				let deposit_witnesses = deposit_witnesses(addresses, &events);
+
+				if !deposit_witnesses.is_empty() {
+					process_call(
+						pallet_cf_ingress_egress::Call::<_, AssethubInstance>::process_deposits {
+							deposit_witnesses,
+							block_height: header.index,
+						}
+						.into(),
+						epoch.index,
+					)
+					.await
+				}
+
+				events
+			}
+		})
+	}
+}
+
+fn address_and_details_to_addresses(
+	address_and_details: Vec<DepositChannelDetails<state_chain_runtime::Runtime, AssethubInstance>>,
+) -> Vec<PolkadotAccountId> {
+	address_and_details
+		.into_iter()
+		.map(|deposit_channel_details| {
+			assert!(
+				deposit_channel_details.deposit_channel.asset == Asset::HubDot ||
+					deposit_channel_details.deposit_channel.asset == Asset::HubUsdc ||
+					deposit_channel_details.deposit_channel.asset == Asset::HubUsdt
+			);
```

Comment:
nit: this assumes we don't add any more hub assets...

If we think we might, we could change this to:

```rust
assert!(hub::Asset::try_from(deposit_channel_details.deposit_channel.asset).is_ok());
```